File: node/tcp-client.js
const raw = require('raw-socket');
const crypto = require('crypto');
const ip = require('ip');

class TCPClient extends raw.Socket {
    constructor(srcIp, srcPort, dstIp, dstPort) {
        super({
            protocol: raw.Protocol.TCP,
            addressFamily: raw.AddressFamily.IPv4
        });

        this.srcIp = srcIp;
        this.srcPort = srcPort;
        this.dstIp = dstIp;
        this.dstPort = dstPort;
    }

    synPacket() {
        // packet length 20 with SYN flag set, other flags are zeroed
        const packet = Buffer.from('0000000000000000000000005002000000000000', 'hex');
        
        packet.writeUInt16BE(this.srcPort, 0, 2);
        packet.writeUInt16BE(this.dstPort, 2, 2);

        // generate four random bytes for the packet sequence
        packet.writeUInt32BE(parseInt(Math.random()*0xffffffff), 4);
    
        // pseudo TCP header
        const pseudoHeader = Buffer.alloc(12);
        pseudoHeader.fill(0);
        pseudoHeader.writeUIntBE(ip.toLong(this.srcIp), 0, 4);
        pseudoHeader.writeUIntBE(ip.toLong(this.dstIp), 4, 4);
        pseudoHeader.writeUIntBE(6, 9, 1); // IP: protocol (ICMP=1, IGMP=2, TCP=6, UDP=17, static value)
        pseudoHeader.writeUIntBE(packet.length, 10, 2);
    
        // calculate checksum
        const checksum = raw.createChecksum(pseudoHeader, packet);
    
        // writing checksum to TCP packet.
        packet.writeUIntBE(checksum, 16, 2);
    
        return packet;
    }
}

module.exports = TCPClient;

File: python/tcpclient.py

import socket
import struct

class TCPClient(socket.socket):
    def __init__(self, srcIp, srcPort, dstIp, dstPort):
        socket.socket.__init__(self, socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_TCP)

        self.srcIp = srcIp
        self.srcPort = srcPort
        self.dstIp = dstIp
        self.dstPort = dstPort

    def synPacket(self):
        # packet length 20 with SYN flag set, other flags are zeroed
        packet =  bytearray(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x50\x02\x00\x00\x00\x00\x00\x00')
        
        struct.pack_into('!H', packet, 0, self.srcPort)
        struct.pack_into('!H', packet, 2, self.dstPort)
    
        # pseudo TCP header
        pseudoHeader = bytearray(b'\x00' * 12)
        struct.pack_into('!I', pseudoHeader, 0, struct.unpack('!I', socket.inet_aton(self.srcIp))[0])
        struct.pack_into('!I', pseudoHeader, 4, struct.unpack('!I', socket.inet_aton(self.dstIp))[0])
        struct.pack_into('!b', pseudoHeader, 9, 6)
        struct.pack_into('!H', pseudoHeader, 10, len(packet))
            
        # calculate checksum
        checksum = TCPClient.__checksum(pseudoHeader, packet)
        # writing checksum to TCP packet.
        struct.pack_into('!H', packet, 16, checksum)
    
        return packet;

    @staticmethod
    def __checksum(header, data):
        sum = 0
        combined = header + data + b'\x00'

        for i in range(0, len(combined) - 1, 2):
            sum += (combined[i] << 8) + combined[i + 1]
            sum  = (sum & 0xffff) + (sum >> 16)

        sum = ~sum & 0xffff
        return sum
